.github/scripts/branch_pr_issue_closer.py

```diff
@@ -135,168 +117,122 @@ def _main_prog():
+        #Create integer list of all open issues:
+        #++++++++++++++++++++++++++++++++++++++
 
-    #Collect all open pull requests:
-    open_pulls = [pr.number for pr in open_repo_pulls]
+        #Extract list of open issues from repo:
+        open_repo_issues = cam_repo.get_issues(state='open')
 
-    #+++++++++++++++++++++++++++++++++++++++++++++++++
-    #Check if one of the keywords exists in PR message
-    #+++++++++++++++++++++++++++++++++++++++++++++++++
+        #Collect all open repo issues:
+        open_issues = [issue.number for issue in open_repo_issues]
 
-    #Keywords are:
-    #close, closes, closed
-    #fix, fixes, fixed
-    #resolve, resolves, resolved
+        #+++++++++++++++++++++++++++++++++++++++++++++
+        #Create integer list of all open pull requests
+        #+++++++++++++++++++++++++++++++++++++++++++++
 
-    #Create regex pattern to find keywords:
-    keyword_pattern = re.compile(r'(^|\s)close(\s|s\s|d\s)|(^|\s)fix(\s|es\s|ed\s)|(^|\s)resolve(\s|s\s|d\s)')
+        #Extract list of open PRs from repo:
+        open_repo_pulls = cam_repo.get_pulls(state='open')
 
-    #Extract (lower case) Pull Request message:
-    pr_msg_lower = merged_pull.body.lower()
+        #Collect all open pull requests:
+        open_pulls = [pr.number for pr in open_repo_pulls]
 
-    #search for at least one keyword:
-    if keyword_pattern.search(pr_msg_lower) is not None:
-        #If at least one keyword is found, then determine location of every keyword instance:
-        word_matches = keyword_pattern.finditer(pr_msg_lower)
-    else:
-        endmsg = "Pull request was merged without using any of the keywords.  Thus there are no issues to close."
-        end_script(endmsg)
+        #+++++++++++++++++++++++++++++++++++++++++++++++++
+        #Check if one of the keywords exists in PR message
+        #+++++++++++++++++++++++++++++++++++++++++++++++++
 
-    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-    #Extract issue and PR numbers associated with found keywords in merged PR message
-    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+        #Compile regex patterns into object:
+        keyword_pattern = re.compile(_LINKED_ISSUE_PATTERN)
 
-    #create issue pattern ("the number symbol {#} + a number"),
-    #which ends with either a space, a comma, a period, or
-    #the end of the string itself:
-    issue_pattern = re.compile(r'#[0-9]+(\s|,|$)|.')
+        #Extract (lower case) Pull Request message:
+        pr_msg_lower = merged_pull.body.lower()
 
-    #Create new "close" issues list:
-    close_issues = []
+        #End script if no keywords found:
+        if keyword_pattern.search(pr_msg_lower) is None:
+            endmsg = f"Pull request #{pr_num} was merged without using any of the keywords. "
+            endmsg += "Thus there are no issues to close."
+            end_script(endmsg)
```

Comment:
Since we are now using the `finditer` function, would it make sense to move this after the for loop and use `if len(close_issues) == 0 and len(close_pulls) == 0)` to avoid scanning the description twice?

Don't suspect this is a hot section of code since it runs on GitHub but if someone writes a book of a description, it might matter?  But if these reads cleaner, please disregard!

Reply:
While this would certainly make the code more performant, the current configuration allows me to have different return messages depending on which step "fails", which I have found handy as sometimes developers come to me telling me the script doesn't work when in fact it was just user error when writing the PR description.  

Of course if you think it would still be worth getting the performance boost then just let me know, as we can always make a new issue and tackle it at a later date.  Thanks!

Reply:
I'm not sure I follow.  The error message and the context would still be the same so that wouldn't change, it's just the location we'd be doing it at, no?

Reply:
Not quite, because that `search` check is to see if you have the correct syntax in the PR description, e. g. `Fixes #123`.  However, after the iteration loop there is a check on the `close_x` lists to see if any of the found issue numbers are actually open, and if not it states that no open issue or PR numbers were found.

In other words it checks that even if your PR description syntax is correct, that the number you are using is valid.  See here:

https://github.com/ESCOMP/CAM-SIMA/blob/development/.github/scripts/branch_pr_issue_closer.py#L253

Hopefully that makes sense?  If not just let me know!

Reply:
Whoops, wrong code quote, I meant start looking here:

https://github.com/ESCOMP/CAM-SIMA/blob/development/.github/scripts/branch_pr_issue_closer.py#L243